File: DailyChallenges/CPP/LeetCode_958.cpp
class Solution {
public:
    bool isCompleteTree(TreeNode* root) {
        queue<TreeNode*> q; q.push(root);
        bool end = false;
        while (!q.empty()){
            int sz = q.size();
            for (int i=0; i<sz; ++i){
                auto node = q.front(); q.pop();
                if (end && node->left) return false;
                if (!end && node->left) q.push(node->left);
                if (!end && !node->left) end = true;
                if (end && node->right) return false;
                if (!end && node->right) q.push(node->right);
                if (!end && !node->right) end = true;     
            }
        }
        return true;
    }
};

File: DailyChallenges/Scala/LeetCode_958.scala
object Solution {
    def isCompleteTree(root: TreeNode): Boolean = {
        val q = collection.mutable.Queue.empty[TreeNode] ; q.enqueue(root)
        var end = false
        while (q.nonEmpty){
            for (i <- 0 until q.length; node = q.dequeue){
                (end, node.left) match {
                    case (false, null) => end = true
                    case (false, _n) => q.enqueue(_n)
                    case (true, _n) if _n != null => return false
                    case _ => null
                }
                (end, node.right) match {
                    case (false, null) => end = true
                    case (false, _n) => q.enqueue(_n)
                    case (true, _n) if _n != null => return false
                    case _ => null
                }
            }
        }
        true
    }
}
